File: lab3/lib/viewmodel/profile_viewmodel.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:flutter/cupertino.dart';

import '../model/user.dart' as my;

class ProfileViewModel extends ChangeNotifier {
  final my.User user;
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseDatabase _database = FirebaseDatabase.instance;

  ProfileViewModel(this.user);

  Future<void> loadData() async {
    final uid = _auth.currentUser?.uid;
    if (uid == null) return;

    final snapshot = await _database.ref().child("users").child(uid).get();
    if (snapshot.exists) {
      final data = snapshot.value as Map<dynamic, dynamic>;
      user.information = data["info"] ?? "";
      user.username = data["name"] ?? "";
      user.surname = data["surname"] ?? "";
      user.patronymic = data["patronymic"] ?? "";
      user.gender = data["gender"] ?? "";
      user.country = data["country"] ?? "";
      user.favoredFilm = data["favoredFilm"] ?? "";
      user.favoredGenre = data["favoredGenre"] ?? "";
      notifyListeners();
    }
  }

  Future<void> update() async {
    final uid = _auth.currentUser?.uid;
    if (uid == null) return;

    final ref = _database.reference().child("users").child(uid);
    final userMap = {
      "login": user.login,
      "email": user.email,
      "info": user.information,
      "name": user.username,
      "surname": user.surname,
      "patronymic": user.patronymic,
      "gender": user.gender,
      "country": user.country,
      "favoredFilm": user.favoredFilm,
      "favoredGenre": user.favoredGenre,
    };
    await ref.set(userMap);
    notifyListeners();
  }

  Future<void> deleteProfile() async {
    final uid = _auth.currentUser?.uid;
    if (uid == null) return;

    await _database.ref().child("users").child(uid).remove();
    await _database.ref().child("favored").child(uid).remove();

    await _auth.currentUser!.delete();
  }

  Future<void> logout() async {
    await _auth.signOut();
  }
}

File: lab1/CoolFilms/CoolFilms/ViewModel/ProfileViewModel.swift
import Foundation
import FirebaseDatabase
import FirebaseAuth
import SwiftUI
import FirebaseStorage

class ProfileViewModel: ObservableObject {
    @Published var user = User(login: "empty", email: "empty", information: "empty", username: "empty", surname: "empty", patronymic: "empty", gender: "empty", country: "empty", favoredFilm: "empty", favoredGenre: "empty")
    
    private let uid = Auth.auth().currentUser?.uid
    private let storage = Storage.storage()
    private let database = Database.database().reference()
    
    private static var cache = NSCache<NSString, UIImage>()
    
    func loadData() {
        let database = Database.database().reference()
        
        database.child("users").child(uid!).observeSingleEvent(of: .value) { (snapshot) in
            if let value = snapshot.value as? [String: Any] {
                self.user.email = value["email"] as? String ?? "empty"
                self.user.login = value["login"] as? String ?? "empty"
                self.user.information = value["information"] as? String ?? "empty"
                self.user.username = value["username"] as? String ?? "empty"
                self.user.surname = value["surname"] as? String ?? "empty"
                self.user.patronymic = value["patronymic"] as? String ?? "empty"
                self.user.gender = value["gender"] as? String ?? "empty"
                self.user.country = value["country"] as? String ?? "empty"
                self.user.favoredFilm = value["favoredFilm"] as? String ?? "empty"
                self.user.favoredGenre = value["favoredGenre"] as? String ?? "empty"
                
                self.objectWillChange.send()
            }
        }
    }
    
    func logOut(){
        do {
            try Auth.auth().signOut()
        } catch let signOutError as NSError {
            print("ERROR")
        }
    }
    
    func deleteProfile(){
        database.child("users").child(uid!).removeValue()
        database.child("favored").child(uid!).removeValue()
        Auth.auth().currentUser?.delete()
    }
    
    func update() {
        let ref = Database.database().reference().child("users").child(uid!)
        let userDictionary: [String: Any] = [
            "login": user.login,
            "country": user.country,
            "gender": user.gender,
            "patronymic": user.patronymic,
            "surname": user.surname,
            "name": user.username,
            "email": user.email,
            "info": user.information,
            "favoredFilm": user.favoredFilm,
            "favoredGenre": user.favoredGenre
        ]
        ref.setValue(userDictionary)
        
        self.objectWillChange.send()
    }
}
